### PR TITLE
Update gitpod-schema.json and add some doc links

### PR DESCRIPTION
**Pull request by axonasif**
## Related Issue(s)

Helps with https://github.com/gitpod-io/gitpod/issues/15731

## Release Notes

```release-note
Add documentation links to Gitpod schema
```

## Build Options:

- [ ] /werft with-github-actions
      Experimental feature to run the build with GitHub Actions (and not in Werft).
- [ ] leeway-no-cache
      leeway-target=components:all
- [ ] /werft no-test
      Run Leeway with `--dont-test`
- [ ] /werft publish-to-npm

#### Preview Environment Options:
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

**Comment by roboquat** [comment]
@axonasif: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

File: components/gitpod-protocol/data/gitpod-schema.json [hunk 1]
```diff
@@ -35,7 +35,7 @@
                             "public"
                         ],
                         "default": "private",
-                        "description": "Whether the port visibility should be private or public. 'private' (default) will only allow users with workspace access to access the port. 'public' will allow everyone with the port URL to access the port."
+                        "description": "Whether the port visibility should be private or public. 'private' (default) will only allow users with workspace access to access the port. 'public' will allow everyone with the port URL to access the port, can help with CORS issues."
```
Comment (geropl) [inline components/gitpod-protocol/data/gitpod-schema.json:38]: Let's drop this change. It's feels too random here.

Would be great if we had a docs page for this, and list the differences in behavior. :cloud:

File: components/gitpod-protocol/data/gitpod-schema.json [hunk 2]
```diff
@@ -71,11 +71,11 @@
                     },
                     "before": {
                         "type": "string",
-                        "description": "A shell command to run before `init` and the main `command`. This command is executed on every start and is expected to terminate. If it fails, the following commands will not be executed."
```
Comment (geropl) [inline components/gitpod-protocol/data/gitpod-schema.json:74]: Why drop this important piece of info? Already helped me multiple times in recalling how the interactions works.

Even better: point to a docs page incl diagram, for instance: https://www.gitpod.io/docs/configure/workspaces/tasks#execution-order

File: components/gitpod-protocol/data/gitpod-schema.json [hunk 3]
```diff
@@ -71,11 +71,11 @@
...
+                        "description": "A shell command to run before `init` and the main `command`."
                     },
                     "init": {
                         "type": "string",
-                        "description": "A shell command to run between `before` and the main `command`. This command is executed only on after initializing a workspace with a fresh clone, but not on restarts and snapshots. This command is expected to terminate. If it fails, the `command` property will not be executed."
```
Comment (geropl) [inline components/gitpod-protocol/data/gitpod-schema.json:78]: Same here: Either leave/update the lifecycle comment, or point to the docs directly: https://www.gitpod.io/docs/configure/workspaces/tasks#execution-order

File: components/gitpod-protocol/data/gitpod-schema.json (hunk 2 above)
Comment (axonasif) [reply components/gitpod-protocol/data/gitpod-schema.json:74]: Thank you for noticing this 😄 

My main concern was this:
> If it fails, the following commands will not be executed.

I assume this is true when you nest `before`, `command`, `init` together in a task?

```yaml
tasks:
  - name: Docs Server
    before: cd docs
    init: npm install
    command: npm run dev
```

I think it's not valid when not nested:

```yaml
tasks:
  - name: random
    before: exit 1

  - name: npm
    command: npm install && npm run dev
```

So, would it be good if we put the docs link instead as you suggested 🙏 ?

File: components/gitpod-protocol/data/gitpod-schema.json (hunk 1 above)
Comment (axonasif) [reply components/gitpod-protocol/data/gitpod-schema.json:38]: > Would be great if we had a docs page for this, and list the differences in behavior. ☁️

Maybe you mean something like this?
- https://github.com/gitpod-io/website/pull/2950

cc @loujaybee 👀

File: components/gitpod-protocol/data/gitpod-schema.json (hunk 3 above)
Comment (axonasif) [reply components/gitpod-protocol/data/gitpod-schema.json:78]: I felt it wasn't fully up-to-date or clear covering all the cases (see [previous comment](https://github.com/gitpod-io/gitpod/pull/16184/files#r1098796318)).

Happy to put the docs link instead, WDYT?

**Comment by akosyakov** [comment]
@loujaybee if you can have a look over it please 🙏

File: components/gitpod-protocol/data/gitpod-schema.json (hunk 2 above)
Comment (svenefftinge) [reply components/gitpod-protocol/data/gitpod-schema.json:74]: We could drop only the last sentence or make it clear that it's about the following items in the array.

Comment (axonasif) [reply components/gitpod-protocol/data/gitpod-schema.json:74]: That also sounds like a good idea.

**Pull request closed without merging**